trie/utils/verkle.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2021 go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/crate-crypto/go-ipa/bandersnatch/fr"
+	"github.com/gballet/go-verkle"
+	"github.com/holiman/uint256"
+)
+
+const (
+	VersionLeafKey    = 0
+	BalanceLeafKey    = 1
+	NonceLeafKey      = 2
+	CodeKeccakLeafKey = 3
+	CodeSizeLeafKey   = 4
+)
+
+var (
+	zero                = uint256.NewInt(0)
+	HeaderStorageOffset = uint256.NewInt(64)
+	CodeOffset          = uint256.NewInt(128)
+	MainStorageOffset   = new(uint256.Int).Lsh(uint256.NewInt(256), 31)
+	VerkleNodeWidth     = uint256.NewInt(256)
+	codeStorageDelta    = uint256.NewInt(0).Sub(CodeOffset, HeaderStorageOffset)
+
+	// BigInt versions of the above.
+	headerStorageOffsetBig = HeaderStorageOffset.ToBig()
+	mainStorageOffsetBig   = MainStorageOffset.ToBig()
+	verkleNodeWidthBig     = VerkleNodeWidth.ToBig()
+	codeStorageDeltaBig    = codeStorageDelta.ToBig()
+
+	getTreePolyIndex0Point *verkle.Point
+)
+
+type PointCache struct {
+	cache map[string]*verkle.Point
```

Comment:
This point cache looks like it leaks a lot of memory, maybe a lru cache is better?

Reply:
yeah we noticed that the memory was growing after a few million blocks, lru is a good idea.